### PR TITLE
Refactor MultiSelect

### PR DESCRIPTION
**Pull request by giancorzo**

### CODE REVIEW

File: source/SIL.AppBuilder.Portal.Frontend/src/ui/components/inputs/multi-select/index.tsx [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+import { attributesFor, relationshipFor } from '@data';
+import { i18nProps } from '@lib/i18n';
+import { Checkbox } from 'semantic-ui-react';
+import ProductIcon from '@ui/components/product-icon';
+
+import { isEmpty } from '@lib/collection';
+
+interface IOwnProps<T> {
+  className?: string;
```
Comment (NullVoxPopuli) [inline source/SIL.AppBuilder.Portal.Frontend/src/ui/components/inputs/multi-select/index.tsx]: If you wanted, you could `& AttributeProps` from `@lib/dom`

File: source/SIL.AppBuilder.Portal.Frontend/src/ui/components/inputs/multi-select/index.tsx [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  list: T[];
+  selected: T[];
+  onChange: (el: T) => void;
+}
+
+export type IProps<T extends ResourceObject> =
+  & IOwnProps<T>
+  & i18nProps;
+
+export class MultiSelect extends React.Component<IProps<ResourceObject>> {
```
Comment (giancorzo) [inline source/SIL.AppBuilder.Portal.Frontend/src/ui/components/inputs/multi-select/index.tsx]: Here is where is was complaining if just use

`export class MultiSelect extends React.Component<IProps<T>>` or 
`export class MultiSelect extends React.Component<IProps>`

File: source/SIL.AppBuilder.Portal.Frontend/src/ui/components/inputs/multi-select/index.tsx [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+      const { data } = relationshipFor(selectedItem, name);
+      return data.id === element.id;
+    });
+
+    return el !== undefined;
+  }
+
+  render() {
+
+    const { list, t } = this.props;
```
Comment (giancorzo) [inline source/SIL.AppBuilder.Portal.Frontend/src/ui/components/inputs/multi-select/index.tsx]: Since the HOC returns different names ("productDefinitions", "stores", "products") how can make it generic

File: source/SIL.AppBuilder.Portal.Frontend/src/ui/components/inputs/multi-select/index.tsx (hunk 2 above)
Comment (NullVoxPopuli) [reply source/SIL.AppBuilder.Portal.Frontend/src/ui/components/inputs/multi-select/index.tsx]: What if you do:

`export class MultiSelect<T> extends React.Component React.Component<IProps<T>> { ...}`
?

File: source/SIL.AppBuilder.Portal.Frontend/src/ui/components/inputs/product-definition-multi-select/index.tsx [hunk 4]
```diff
@@ -4,10 +4,11 @@ import { withTranslations } from '@lib/i18n';
...
+//import { Display } from './display';
+import { MultiSelect } from '@ui/components/inputs/multi-select';
 
 export default compose(
   withProductDefinitions(),
   withLoader(({ error, productDefinitions }) => !error && !productDefinitions),
   withTranslations,
-)(Display);
\ No newline at end of file
+)(MultiSelect);
```
Comment (giancorzo) [inline source/SIL.AppBuilder.Portal.Frontend/src/ui/components/inputs/product-definition-multi-select/index.tsx]: how can I define what type is the component going to use?

File: source/SIL.AppBuilder.Portal.Frontend/src/ui/components/inputs/multi-select/index.tsx (hunk 2 above)
Comment (giancorzo) [reply source/SIL.AppBuilder.Portal.Frontend/src/ui/components/inputs/multi-select/index.tsx]: This one works better

`export type IProps<T> =
  & IOwnProps<T>
  & i18nProps;

export class MultiSelect<T extends ResourceObject> extends React.Component<IProps<T>> {`

File: source/SIL.AppBuilder.Portal.Frontend/src/ui/components/inputs/multi-select/index.tsx (hunk 3 above)
Comment (NullVoxPopuli) [reply source/SIL.AppBuilder.Portal.Frontend/src/ui/components/inputs/multi-select/index.tsx]: pass `list={productDefinition}`?

I don't understand the issue. :-\

if the MultiSelect is for some reason in a HoC mountain, you could use recompose`s withProps: https://github.com/acdlite/recompose/blob/master/docs/API.md#withprops

File: source/SIL.AppBuilder.Portal.Frontend/src/ui/components/inputs/product-definition-multi-select/index.tsx (hunk 4 above)
Comment (NullVoxPopuli) [reply source/SIL.AppBuilder.Portal.Frontend/src/ui/components/inputs/product-definition-multi-select/index.tsx]: ah ok, I understand now -- https://github.com/acdlite/recompose/blob/master/docs/API.md#withprops

Comment (giancorzo) [reply source/SIL.AppBuilder.Portal.Frontend/src/ui/components/inputs/product-definition-multi-select/index.tsx]: ok got it

File: source/SIL.AppBuilder.Portal.Frontend/src/ui/components/inputs/multi-select/index.tsx [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  selected: T[];
+  onChange: (el: T) => void;
+  emptyListLabel: string;
+}
+
+export class MultiSelect<T extends ResourceObject> extends React.Component<IOwnProps<T>> {
+
+  onChange = (element) => (e) => {
+    e.preventDefault();
+    this.props.onChange(element);
```
Comment (NullVoxPopuli) [inline source/SIL.AppBuilder.Portal.Frontend/src/ui/components/inputs/multi-select/index.tsx]: I think we need tests, because this should fail the build

File: source/SIL.AppBuilder.Portal.Frontend/src/ui/components/inputs/multi-select/index.tsx [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+      const { data } = relationshipFor(selectedItem, relationshipName);
+      return data.id === element.id;
+    });
+
+    return el !== undefined;
+  }
+
+  render() {
+
+    const { list, emptyListLabel } = this.props;
```
Comment (NullVoxPopuli) [inline source/SIL.AppBuilder.Portal.Frontend/src/ui/components/inputs/multi-select/index.tsx]: API-wise emptyListLabel should prob not be required, but overridable

File: source/SIL.AppBuilder.Portal.Frontend/src/ui/components/inputs/multi-select/index.tsx [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+            data-test-item
+            onClick={this.onChange(element)}
+          >
+            <Checkbox
+              data-test-item-checkbox
+              className='m-r-md'
+              value={element.id}
+              checked={this.inSelectedList(element)}
+            />
+            <ProductIcon product={element} />
```
Comment (NullVoxPopuli) [inline source/SIL.AppBuilder.Portal.Frontend/src/ui/components/inputs/multi-select/index.tsx]: are all usages of this component going to use a product icon?

File: source/SIL.AppBuilder.Portal.Frontend/src/ui/components/inputs/product-definition-multi-select/index.tsx [hunk 8]
```diff
@@ -1,13 +1,18 @@
...
 export default compose(
+  withTranslations,
   withProductDefinitions(),
   withLoader(({ error, productDefinitions }) => !error && !productDefinitions),
-  withTranslations,
-)(Display);
\ No newline at end of file
+  withProps(({ productDefinitions, t }) => ({
+    list: productDefinitions,
+    relationshipName: 'productDefinition',
```
Comment (NullVoxPopuli) [inline source/SIL.AppBuilder.Portal.Frontend/src/ui/components/inputs/product-definition-multi-select/index.tsx]: why is this needed?

File: source/SIL.AppBuilder.Portal.Frontend/src/ui/components/inputs/store-multi-select/index.tsx [hunk 9]
```diff
@@ -1,13 +1,18 @@
...
 export default compose(
+  withTranslations,
   withStores(),
   withLoader(({ error, stores }) => !error && !stores),
-  withTranslations,
-)(Display);
\ No newline at end of file
+  withProps(({ stores, t }) => ({
+    list: stores,
+    relationshipName: 'store',
```
Comment (NullVoxPopuli) [inline source/SIL.AppBuilder.Portal.Frontend/src/ui/components/inputs/store-multi-select/index.tsx]: relation to what?

File: source/SIL.AppBuilder.Portal.Frontend/src/ui/components/inputs/multi-select/index.tsx (hunk 7 above)
Comment (giancorzo) [reply source/SIL.AppBuilder.Portal.Frontend/src/ui/components/inputs/multi-select/index.tsx]: at least the one related to product yes, for now, there is no requirement for conditional

File: source/SIL.AppBuilder.Portal.Frontend/src/ui/components/inputs/product-definition-multi-select/index.tsx (hunk 8 above)
Comment (giancorzo) [reply source/SIL.AppBuilder.Portal.Frontend/src/ui/components/inputs/product-definition-multi-select/index.tsx]: it's needed because I need to 
`const { data } = relationshipFor(selectedItem, relationshipName);`

File: source/SIL.AppBuilder.Portal.Frontend/src/ui/components/inputs/store-multi-select/index.tsx (hunk 9 above)
Comment (giancorzo) [reply source/SIL.AppBuilder.Portal.Frontend/src/ui/components/inputs/store-multi-select/index.tsx]: it's needed because I need to
`const { data } = relationshipFor(selectedItem, relationshipName);`

File: source/SIL.AppBuilder.Portal.Frontend/src/ui/components/inputs/product-definition-multi-select/index.tsx (hunk 8 above)
Comment (NullVoxPopuli) [reply source/SIL.AppBuilder.Portal.Frontend/src/ui/components/inputs/product-definition-multi-select/index.tsx]: but what does it semantically represent
